src/ejabberd_tls.erl

```diff
@@ -130,10 +131,19 @@ get_peer_certificate(#ejabberd_tls_socket{tls_module = fast_tls, tls_socket = S,
 -spec close(socket()) -> ok.
 close(#ejabberd_tls_socket{tls_module = M, tls_socket = S}) -> M:close(S).
 
-
 -spec get_sockmod(socket()) -> module().
 get_sockmod(#ejabberd_tls_socket{tls_module = Module}) -> Module.
 
+-spec get_tls_last_message(peer | owner, ejabberd_socket:socket()) -> {ok, binary()} | binary().
```

Comment:
I actually don't like that return value API. Sometimes it returns a tagged `ok` tuple with the real data, sometimes an empty binary, and also sometimes a tagged `error` tuple from the fast_tls layer directly. I'd say the return values should be like `{ok, binary()}` for the case that contains the actual correct data, and `{error, term()}`, where term could be an atom like `no_finished_message_yet` or `no_ssl_connection_so_impossible_to_have_a_finished_message_you_idiot` or `undefined` or whatever appropiate things should be given, atoms or strings or whatever 😛
In escalus I did it as https://github.com/esl/escalus/blob/807f660d344db23e820fcf896d4756f05d993a6c/src/escalus_tcp.erl#L260-L265 for example. And in fast_tls, [it will return ](https://github.com/processone/fast_tls/blob/536a9893b2e7d699bee1c2b567f0765bb7849015/c_src/fast_tls.c#L1231)`{error, undefined}` if the message was empty.

That could also help cleaning that `cyrsasl_scram:maybe_get_tls_last_message/2` helper function.

Reply:
Also, fast_tls is expecting `peer | self`, not `peer | owner`. At least that's how I made the PR there and it's already merged so... 🤷‍♂️